pkg/invoke/args.go

```diff
@@ -80,3 +82,23 @@ func stringify(pluginArgs [][2]string) string {
 
 	return strings.Join(entries, ";")
 }
+
+// DelegateArgs implements the CNIArgs interface
+// used for delegation to inherit from environments
+// and allow some overrides like CNI_COMMAND
+var _ CNIArgs = &DelegateArgs{}
+
+type DelegateArgs struct {
+	Command string
+}
+
+func (d *DelegateArgs) AsEnv() []string {
+	env := os.Environ()
+
+	// The custom values should come in the end to override the existing
+	// process environment of the same key.
+	env = append(env,
+		"CNI_COMMAND="+d.Command,
+	)
+	return env
+}
```

Comment:
Wouldn't it be simpler to just do:

```
diff --git a/pkg/invoke/delegate.go b/pkg/invoke/delegate.go
index 30b4672f11fc..785f63b501dd 100644
--- a/pkg/invoke/delegate.go
+++ b/pkg/invoke/delegate.go
@@ -44,6 +44,13 @@ func delegateCommon(expectedCommand, delegatePlugin string, exec Exec) (string,
 // DelegateAdd calls the given delegate plugin with the CNI ADD action and
 // JSON configuration
 func DelegateAdd(ctx context.Context, delegatePlugin string, netconf []byte, exec Exec) (types.Result, error) {
+       cmd := os.Getenv("CNI_COMMAND")
+       if cmd == "" {
+               return fmt.Errorf("environment variable CNI_COMMAND must be specified.")
+       }
+       // restore CNI_COMMAND to original value upon return.
+       defer os.Setenv("CNI_COMMAND", cmd)
+       os.Setenv("CNI_COMMAND", "ADD")
        pluginPath, realExec, err := delegateCommon("ADD", delegatePlugin, exec)
        if err != nil {
                return nil, err
@@ -55,6 +62,13 @@ func DelegateAdd(ctx context.Context, delegatePlugin string, netconf []byte, exe
 // DelegateCheck calls the given delegate plugin with the CNI CHECK action and
 // JSON configuration
 func DelegateCheck(ctx context.Context, delegatePlugin string, netconf []byte, exec Exec) error {
+       cmd := os.Getenv("CNI_COMMAND")
+       if cmd == "" {
+               return fmt.Errorf("environment variable CNI_COMMAND must be specified.")
+       }
+       // restore CNI_COMMAND to original value upon return.
+       defer os.Setenv("CNI_COMMAND", cmd)
+       os.Setenv("CNI_COMMAND", "CHECK")
        pluginPath, realExec, err := delegateCommon("CHECK", delegatePlugin, exec)
        if err != nil {
                return err
@@ -66,6 +80,13 @@ func DelegateCheck(ctx context.Context, delegatePlugin string, netconf []byte, e
 // DelegateDel calls the given delegate plugin with the CNI DEL action and
 // JSON configuration
 func DelegateDel(ctx context.Context, delegatePlugin string, netconf []byte, exec Exec) error {
+       cmd := os.Getenv("CNI_COMMAND")
+       if cmd == "" {
+               return fmt.Errorf("environment variable CNI_COMMAND must be specified.")
+       }
+       // restore CNI_COMMAND to original value upon return.
+       defer os.Setenv("CNI_COMMAND", cmd)
+       os.Setenv("CNI_COMMAND", "DEL")
        pluginPath, realExec, err := delegateCommon("DEL", delegatePlugin, exec)
        if err != nil {
                return err
```

Reply:
@mars1024 @mccv1r0 yeah, that's what I would have done here.

Reply:
@dcbw @mccv1r0  I think `DelegateArgs` seems more proper than `set/unset env`because ...
1. It's centralized, we don't have to do code change in **every** delegate funcs
2. It can be easily expanded, maybe in the future, we need to override another env ?
3. It has no influence on main process' env because we don't change it but just override it in heritance, this maybe less risky

But I also think the simpler way which you talk about can work, if you cloud persuade me about the three points above, I'd be happy to change the code ~ thanks ~

Reply:
hi, @dcbw @mccv1r0 , any further suggestions for promoting this? I'm very happy to change the code if needed.